app/models/metadatum.rb

```diff
@@ -0,0 +1,101 @@
+class Metadatum < ApplicationRecord
+  belongs_to :sample
+
+  # Types: 0 for string, 1 for number, 2 for lists of values (as string)
+  KEY_TO_TYPE = { unique_id: 0, sample_type: 0, nucleotide_type: 0, collection_date: 0, collection_location: 0, collected_by: 0, age: 1, gender: 0, race: 0, primary_diagnosis: 0, antibiotic_administered: 0, admission_date: 0, admission_type: 0, discharge_date: 0, discharge_type: 0, immunocomp: 0, other_infections: 0, comorbitity: 0, known_organism: 0, infection_class: 0, detection_method: 0, library_prep: 0, sequencer: 0, rna_dna_input: 1, library_prep_batch: 0, extraction_batch: 0 }.freeze
+
+  KEY_TO_STRING_OPTIONS = {
+    nucleotide_type: %w[DNA RNA],
+    gender: %w[Female Male],
+    race: ["Caucasian", "Asian", "African American", "Other"],
+    admission_type: %w[ICU General],
+    discharge_type: ["ICU", "Hospital", "30 Day Mortality", "Other"],
+    infection_class: ["Definite", "No Infection", "Suspected", "Unknown", "Water Control"],
+    library_prep: ["NEB Ultra II FS DNA", "NEB RNA Ultra II", "NEB Ultra II Directional RNA", "NEB Utra II DNA", "Nextera DNA", "Other"],
+    sequencer: %w[MiSeq NextSeq HiSeq NovaSeq Other]
+  }.freeze
+
+  # Add a new Metadatum entry to the sample
+  def self.add_to_sample(sample, key, raw_value)
+    key = key.to_sym
+
+    # validate raises RuntimeError if invalid. Caller can handle and provide
+    # feedback.
+    validated_value = validate(key, raw_value)
```

Comment:
I think you should look into rail-style validations e.g. the validates_with hook (https://guides.rubyonrails.org/active_record_validations.html#validates-with). That way you could call a standard create method and the validations would be carried out automatically

Reply:
There’s no conflict. When inserting a single metadatum, you would call a standard create method with a validates_with referring to your custom validation function. When inserting multiple metadata, you would run your

Reply:
There’s no conflict. When inserting a single metadatum, you would call a standard create method with a validates_with referring to your custom validation function. When inserting multiple metadata, you would run your

Reply:
Coming from rails ignorance here: would you have to disabled validation to do the bulk insert? Or were you saying to bypass the model?

Reply:
I meant bypass the model -- I don't think there is a standard rails way to create objects in bulk.
However:
(a) There is a gem that lets you create in bulk through ActiveRecord, although I don't know if we would want to use it. See Joao's answer here: https://stackoverflow.com/questions/39064067/activerecord-skip-validation-when-saving-multiple-objects). To be tested.
(b) Another option I had never heard of would be to wrap all the saves in a transaction to make it faster -- see Harish's answer here: https://stackoverflow.com/questions/2509320/saving-multiple-objects-in-a-single-call-in-rails. Sounds like it is not as good as a single query, but much better than without the transaction. To be tested.

Reply:
i am not sure about the rails validate. my understanding is that if validates fail, the active record will fail to save. Since we are doing a softcheck here, I am not sure if it;s necessary. The validation context here is quite different from the rails validation context i think

Reply:
also instead of calling validate directly,  add a before_save hook to calculate validated value.